src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php

```diff
@@ -39,57 +39,59 @@ protected function processValue($value, $isRoot = false)
     }
 
-    /**
-     * Populates the class and tags from parent definitions.
-     */
-    private function resolveDefinition(ChildDefinition $definition)
+    private function mergeInstanceofDefinition(Definition $def, ChildDefinition $instanceofDefinition)
     {
-        if (!$this->container->has($parent = $definition->getParent())) {
-            return;
+        $configured = $def->getChanges();
+        $changes = $instanceofDefinition->getChanges();
+        if (!isset($configured['shared']) && isset($changes['shared'])) {
+            $def->setShared($instanceofDefinition->isShared());
         }
-
-        $parentDef = $this->container->findDefinition($parent);
-        $class = $parentDef instanceof ChildDefinition ? $this->resolveDefinition($parentDef) : $parentDef->getClass();
-        $class = $definition->getClass() ?: $class;
-
-        // append parent tags when inheriting is enabled
-        if ($definition->getInheritTags()) {
-            $definition->setInheritTags(false);
-
-            foreach ($parentDef->getTags() as $k => $v) {
-                foreach ($v as $v) {
-                    $definition->addTag($k, $v);
-                }
-            }
+        if (!isset($configured['configurator']) && isset($changes['configurator'])) {
+            $def->setConfigurator($instanceofDefinition->getConfigurator());
         }
-
-        return $class;
-    }
-
-    private function mergeDefinition(Definition $def, ChildDefinition $definition)
-    {
-        $changes = $definition->getChanges();
-        if (isset($changes['shared'])) {
-            $def->setShared($definition->isShared());
+        if (!isset($configured['public']) && isset($changes['public'])) {
+            $def->setPublic($instanceofDefinition->isPublic());
         }
-        if (isset($changes['abstract'])) {
-            $def->setAbstract($definition->isAbstract());
+        if (!isset($configured['lazy']) && isset($changes['lazy'])) {
+            $def->setLazy($instanceofDefinition->isLazy());
         }
-
-        ResolveDefinitionTemplatesPass::mergeDefinition($def, $definition);
-
-        // prepend instanceof tags
-        $tailTags = $def->getTags();
-        if ($headTags = $definition->getTags()) {
-            $def->setTags($headTags);
-            foreach ($tailTags as $k => $v) {
+        if (!isset($configured['autowired']) && isset($changes['autowired'])) {
+            $def->setAutowired($instanceofDefinition->getAutowired());
+        }
+        // merge properties
+        $properties = $def->getProperties();
+        foreach ($instanceofDefinition->getProperties() as $k => $v) {
+            // don't override properties set explicitly on the service
+            if (!isset($properties[$k])) {
+                $def->setProperty($k, $v);
+            }
+        }
+        // merge method calls
+        if ($instanceofCalls = $instanceofDefinition->getMethodCalls()) {
+            $currentCallMethods = array_map(function($call) {
```

Comment:
Ah, but now I also need to `strtolower` for the case insensitive matching later :)